### PR TITLE
qmail-remote: avoid recoding CRLF to CRCRLF

### PR DESCRIPTION
**Pull request by DerDakon**
If qmail-remote get's a '\r' (CR, carriage return) in it's input stream (which means that a mail injected into qmail-queue contained a '\r') it will send this unmasked to the network. This patch works this way: if there is a single '\r' somewhere in the input stream it is handled as a line break, so it is send as CRLF to the network. If there is a CRLF sequence in the input stream (normally there should only be bare LFs) it will not double the CR (like qmail-remote normally does) but send this CRLF sequence as is.

### CODE REVIEW

**Comment by mbhangui** [comment]
The treatment for bare LF is fine but I think better to leave bare CR unmasked. I think, this will break DKIM and the signature computed with the bare CR will no longer match when the mail body has the bare CR corrected.

**Comment by DerDakon** [comment]
It's a bit of a grey area, IIRC the SMTP RfCs do no not allow either CR or LF to show up without the other. It's just garbage in, garbage out.

**Comment by schmonz** [comment]
@mbhangui, you're doing DKIM in the wild, right? I'd love to know whether this change breaks anything so we can weigh benefits against costs with our eyes open.

**Comment by mbhangui** [comment]
> @mbhangui, you're doing DKIM in the wild, right? I'd love to know whether this change breaks anything so we can weigh benefits against costs with our eyes open.
In real life I haven't encountered mails with bare CR. It could be possible if a mail is being injected and the body contains bare CR. If we do dkim signing before it gets passed to qmail-remote, the DKIM signature computed on the recipient host will not match. So the only case where this will break is

1. If we are injecting emails on a system using qmail-inject and the body contains bare CR.
2. We do DKIM Signing before the mail gets into the queue (e.g. a DKIM qmail-queue replacement) or if we do DKIM signing just before the mail content gets passed to qmail-remote (e.g. using QMAILREMOTE feature).

So we are talking of a case where we have improper mail content (with bare CR) that is passed to qmail-inject.

For all valid, proper mail content, this issue will not happen.

**Comment by mbhangui** [comment]
If we do DKIM signing and we convert the bare CR to CRLF in the DKIM signing program, then this issue will not happen. So when we implement DKIM signing, we should have the same code in the dkim signing process.

**Comment by schmonz** [comment]
I'm tentatively convinced this is worth the risk. I'd be more convinced if we knew of real-world cases where a message with bare CRs had gone out on the wire (not involving DKIM at all) and other SMTP servers rejected or dropped or garbled it.

Would like to hear at least one more person's opinion on this before we merge.

**Comment by DerDakon** [comment]
I have seen this happening when programs (mostly PHP scripts) injects mails with CRLF line endings instead of just LF line endings. You can try yourself doing

```bash
echo -e 'Subject: CRLFTest\n\ncr\r\nlf\n' | qmail-inject -a remote@mail.address
```

And then use wireshark or tcpdump and see that CRCRLF sequences go over the wire, which is illegal.

**Comment by DerDakon** [comment]
Sorry, accidentially deleted wrong branch.

File: qmail-remote.c
Comment (schmonz) [inline qmail-remote.c:206]: Looks like tab instead of 2-space indent?

File: qmail-remote.c
Comment (schmonz) [inline qmail-remote.c:210]: Missing a `{`? If not, this looks odd.

Comment (DerDakon) [reply qmail-remote.c:210]: It's in line 203.

**Review by schmonz — APPROVED** [review]
Thank you for putting in the effort to prepare these tests, and for your patience waiting for my review.

**Pull request merged**
